src/openfermion/hamiltonians/_general_hubbard.py

```diff
@@ -0,0 +1,288 @@
+        - :math:`t_{a, b}^{(\mathrm{nghbr})}` is the tunneling amplitude
+          between spin orbitals on neighboring sites,
+        - :math:`U_{a, b}^{(\mathrm{onsite, \pm})}` is the Coulomb potential
+          between spin orbitals on the same site with the same (+) or different
+          (-) spins,
+        - :math:`U_{a, b}^{(\mathrm{nghbr, \pm})}` is the Coulomb potential
+          betwen spin orbitals on neighborings sites with the same (+) or
+          different (-) spins,
+        - :math:`\mu_{a}` is the chemical potential.
+
+    One can also construct the Hamiltonian for the spinless model, which
+    has the form
+
+    .. math::
+
+        \begin{align}
+        H = &- \sum_{a < b} t_{a, b}^{(\mathrm{onsite})}
+               \sum_{i}
+                     (a^\dagger_{i, a} a_{i, b} +
+                      a^\dagger_{i, b} a_{i, a})
+            \\
+            &- \sum_{a} t_{a, a}^{(\mathrm{nghbr})}
+               \sum_{\{i, j\}}
+                     (a^\dagger_{i, a} a_{j, a} +
+                      a^\dagger_{j, a} a_{i, a})
+             - \sum_{a < b} t_{a, b}^{(\mathrm{nghbr})}
+               \sum_{(i, j)}
+                     (a^\dagger_{i, a} a_{j, b} +
+                      a^\dagger_{j, b} a_{i, a})
+            \\
+            &+ \sum_{a < b} U_{a, b}^{(\mathrm{onsite})}
+               \sum_{i}
+                     n_{i, a} n_{i, b}
+            \\
+            &+ \sum_{a} U_{a, a}^{(\mathrm{nghbr})}
+               \sum_{\{i, j\}}
+                     n_{i, a} n_{j, a}
+             + \sum_{a < b} U_{a, b}^{(\mathrm{nghbr})}
+               \sum_{(i, j)}
+                     n_{i, a} n_{j, b}
+            \\
+            &- \sum_{a} \mu_a
+               \sum_i n_{i, a}
+        \end{align}
+    """
+
+
+    def __init__(self, lattice,
+                 tunneling_parameters=None,
+                 interaction_parameters=None,
+                 potential_parameters=None
+                 ):
+        """A Hubbard model defined on a lattice.
+
+        Args:
+            lattice (HubbardLattice): The lattice on which the model is defined.
+            tunneling_parameters (Iterable[Tuple[Hashable, Tuple[int, int],
+                Number]], optional): The tunneling parameters.
+            interaction_parameters (Iterable[Tuple[Hashable, Tuple[int, int],
+                Number, int?]], optional): The interaction parameters.
+            potential_parameters (Iterable[Tuple[int, Number]], optional): The
+                potential parameters.
+
+        Each group of parameters is specified as an iterable of tuples.
+
+        Each tunneling parameter is a tuple (edge_type, dofs, coefficient).
+
+        Each interaction parameter is a tuple (edge_type, dofs,
+        coefficient, spin_pairs). The final spin_pairs element is
+        optional, and will default to SpinPairs.ALL. In any case, it is
+        ignored for spinless lattices.
+
+        Each potential parameter is a tuple (dof, coefficient).
```

Comment:
Please explain "dof".

Reply:
While you're at it, please explain where `edge_type` comes from. This docstring needs to be expanded to make it easier for the user to know how to form the arguments.